File: 2019141460416/my program.py
# Homework
计算机网络03班课程大作业
# *** python 多线程web服务器的设计    ****#
# *************************************#
from socket import *
'''引入线程模块'''
import threading
import os

'''自定义线程函数'''
'''针对线程使用编写的函数'''
def Server(tcpClisock, addr):
    '''将缓冲区大小设置为1KB'''
    BUFSIZE = 1024
    print('Waiting for the connection：', addr)
    data = tcpClisock.recv(BUFSIZE).decode()
    filename = data.split()[1]
    filename = filename[1:]

    '''当网络质量差没有收到浏览器的访问数据时执行'''
    if filename == "":
        tcpClisock.close()
        print("请输入要访问的文件")

    base_dir = os.getcwd()
    file_dir = os.path.join(base_dir,filename)

    '''当访问的文件在本地服务器存在时执行'''
    if os.path.exists(file_dir):
        f = open(file_dir,encoding = 'utf-8')
        '''构造成功报文反馈给服务器'''
        SUCCESS_PAGE = "HTTP/1.1 200 OK\r\n\r\n" + f.read()
        print(SUCCESS_PAGE)
        '''返回给客户端浏览器成功的页面'''
        tcpClisock.sendall(SUCCESS_PAGE.encode())
        '''关闭专门针对一个客户机程序创建的新套接字'''
        tcpClisock.close()
    else:
        '''如果不存在，返回失败页面'''
        FAIL_PAGE = "HTTP/1.1 404 NotFound\r\n\r\n" + open(os.path.join(base_dir, "fail.html"), encoding="utf-8").read()
        print(FAIL_PAGE)
        '''返回给客户端浏览器失败的页面，sendall函数只可以发送字节类型，对字符串数据进行转换'''
        tcpClisock.sendall(FAIL_PAGE.encode())
        '''关闭专门针对一个客户机程序创建的新套接字'''
        tcpClisock.close()

'''主函数'''
if __name__ == '__main__':

    '''分配IP、端口、创建套接字对象'''
    '''地址=主机名+端口号'''
    ADDR = ("", 8080)
    '''创建套接字对象'''
    tcpSersock = socket(AF_INET, SOCK_STREAM)
    '''将套接字绑定到服务器地址，且绑定的是源ip地址以及源端口号。始终欢迎别的套接字的申请接入。'''
    tcpSersock.bind(ADDR)
    '''启用服务器监听,此处说明在连接被转接或拒绝之前允许传入最大链接请求数为5'''
    tcpSersock.listen(5)
    print("waiting for connection......\n")
    '''进入无限循环'''
    while True:
        '''调用 accept()函数开启一个单线程服务器，等待客户端连接'''
        tcpClisock, addr = tcpSersock.accept()
        thread = threading.Thread(target=Server, args=(tcpClisock, addr))
        '''开始执行该线程'''
        thread.start()
    tcpSersock.close()

File: 2019141090145/myprogram.cpp
from socket import *
import sys
import os
import time
serverName = 'DESKTOP-RPMTUJK'
serverPort = 12003

#if len(sys.argv) <= 1:
#    print ('Usage : "python ProxyServer.py server_ip"\n[server_ip : It is the IP Address Of Proxy Server')
#	sys.exit(2)

# 为代理服务器 创建一个TCP套接字、绑定端口号、设置服务器最大连接客户机数量为3(因为是多线程Web服务器)
tcpSerSock = socket(AF_INET, SOCK_STREAM)
tcpSerSock.bind(('',1234))
tcpSerSock.listen(3)

while 1:
	# 准备从客户机接收响应消息
	print ('准备从客户机接收响应消息...')
	tcpCliSock, addr = tcpSerSock.accept()
	print ('接收到一个连接，来自：', addr)
	
	# 获取客户机发送过来的消息
	message = tcpCliSock.recv(4096).decode()
	print ('客户机发送过来的消息：',message)
	
	# 从消息从提取出文件名
	filename = message.split()[1].partition("//")[2].replace('/','_')
	print ('文件名：',filename)
	fileExist = "false"
	
	try:
		# 检查要访问的文件是否在此Web代理服务器中
		print ('开始检查代理服务器中是否存在文件：',filename)
		f = open(filename, "r")
		outputdata = f.readlines()
		fileExist = "true"
		print ('文件存在在代理服务器中')
		# 文件存在在代理服务器中，返回响应消息(请求的web网页)给客户机
		#tcpCliSock.send("HTTP/1.0 200 OK\r\n")
		#tcpCliSock.send("Content-Type:text/html\r\n")
		for i in range(0,len(outputdata)):
				tcpCliSock.send(outputdata[i].encode())	
		print ('Read from cache')
			
	# 文件不在代理服务器当中，代理服务器就会向远端服务器请求消息，保存好了再返回给客户机
	except IOError:
		if fileExist == "false":
			print ('文件不在代理服务器当中，开始向远端服务器请求网页')
			# 在代理服务器中创建一个TCP套接字
			c = socket(AF_INET,SOCK_STREAM)
			
			hostn = message.split()[1].partition("//")[2].partition("/")[0]
			print ('Host Name: ',hostn)
			try:
				# TCP套接字c 连接到远端服务器80端口
				c.connect((hostn,80))
				print ('套接字连接到主机的80号端口')
				
				# 在套接字上创建一个临时的文件，而且要向80端口(远端服务器)请求信息
				#for the file requested by the client
				#fileobj = c.makefile('r', 0)
				#fileobj.write("GET "+"http://" + filename + " HTTP/1.0\n\n")
				
				# 代理服务器 读取 从远端服务器从响应的消息
				c.sendall(message.encode())
				buff = c.recv(4096)
				tcpCliSock.sendall(buff)
				
				# 在代理服务器中 创建一个新的文件 用于存放请求过来的消息
				# 将代理服务器中的响应 发送到客户端套接字，并将相应的文件发送到缓存中
				tmpFile = open("./" + filename,"w")
				tmpFile.writelines(buff.decode().replace('\r\n','\n'))
				tmpFile.close()
				
			except:
				print ("代理服务器向远端服务器请求网页失败")
		else:
			# 如果客户机请求的消息在远端服务器也找不到，就说明请求不到了
			print ('文件存在，但是还是出现了 IOError异常')

	# 关闭客户机 和 代理服务器的TCP套接字
	print ('关闭套接字：tcpCliSock')
	tcpCliSock.close()
# 关闭代理服务器 和 服务器的TCP套接字
print ('关闭套接字：tcpSerSock')
tcpSerSock.close()
